### ISSUE TITLE
stdin.pause/resume not working if there is a pause

### ISSUE THREAD

**Issue by jackcannon** [oven-sh/bun#8693]
### What version of Bun is running?

1.0.26+c75e768a6

### What platform is your computer?

Darwin 23.1.0 x86_64 i386

### What steps can reproduce the bug?

I have made a small demo repo that illustrates the problem: https://github.com/jackcannon/stdin-bug-bun/tree/master

* `bun run example:node` - ✅ Works - runs the script using node
* `bun run example:bun` - ❌ Broken - runs the script using bun
* `bun run example:bun:ignore` - ✅ Works - runs the script using bun WITHOUT pausing between the prompts

The script it runs is: https://github.com/jackcannon/stdin-bug-bun/blob/master/stdin.js

Output (GIF):
![Kapture 2024-02-05 at 16 48 45](https://github.com/oven-sh/bun/assets/1922422/97b25de2-1aa1-402c-bab6-b404db0a9c93)

Output (text):
```
[16:47:51] stdin-bug-bun (master)$ bun --version
1.0.26
[16:48:11] stdin-bug-bun (master)$ bun run example:node
$ node stdin.js
What is your name - a? › jack
Pausing for 1 second...
What is your name - b? › jackb

RESULT: { a: 'jack', b: 'jackb' }
[16:48:23] stdin-bug-bun (master)$ bun run example:bun
$ bun run stdin.js
What is your name - a? › jack
Pausing for 1 second...
What is your name - b? ›

TIMEOUT - exiting...
[16:48:34] stdin-bug-bun (master)$ ackb
-bash: ackb: command not found
[16:48:34] stdin-bug-bun (master)$ bun run example:bun:ignore
$ bun run stdin.js --ignore-pause
What is your name - a? › jack
Skipping pause...
What is your name - b? › jackb

RESULT: {
  a: "jack",
  b: "jackb",
}
[16:48:41] stdin-bug-bun (master)$
```

The script consists of 2 'prompts' with a setTimeout between them. Each prompt does this at the start:
```typescript
      process.stdin.setRawMode(true);
      process.stdin.resume();
      process.stdin.on('data', listenFn);
      process.stdin.setEncoding('utf8');
```

and this when the return/enter button is registered:
```typescript
      process.stdin.setRawMode(false);
      process.stdin.pause();
      process.stdin.off('data', listenFn);
      resolve(result); // resolves the promise - continuing the script
```

In node, this works. Both prompts allow for input to be listened to.
However, in bun, the second prompt does not receive any data if there is a pause between the prompts. If there is no pause, both receive data fine.

It appears to be related to `process.stdin.pause`/`process.stdin.resume`, as when I remove the `process.stdin.pause()` in the `stop` function, it fixes this issue. *HOWEVER*, this prevents the script from exiting when it's finished (see https://stackoverflow.com/questions/59220095/node-doesnt-exit-automatically-once-a-listener-is-set-on-stdin )

**Note**: there is a `IS_DEBUG` flag in the code, that logs out some stats at the beginning and end of both the `start` and `end` functions. With this enabled, the corresponding output is:

```
[16:43:08] stdin-bug-bun (master)$ bun run example:node
$ node stdin.js
BEFORE - start { isRaw: false, isPaused: false, encoding: null }
AFTER  - start { isRaw: true, isPaused: false, encoding: 'utf8' }
What is your name - a? › jack
BEFORE - stop  { isRaw: true, isPaused: false, encoding: 'utf8' }
AFTER  - stop  { isRaw: false, isPaused: true, encoding: 'utf8' }

Pausing for 1 second...
BEFORE - start { isRaw: false, isPaused: true, encoding: 'utf8' }
AFTER  - start { isRaw: true, isPaused: false, encoding: 'utf8' }
What is your name - b? › jackb
BEFORE - stop  { isRaw: true, isPaused: false, encoding: 'utf8' }
AFTER  - stop  { isRaw: false, isPaused: true, encoding: 'utf8' }


RESULT: { a: 'jack', b: 'jackb' }
[16:43:27] stdin-bug-bun (master)$ bun run example:bun
$ bun run stdin.js
BEFORE - start { isRaw: false, isPaused: false, encoding: null }
AFTER  - start { isRaw: true, isPaused: false, encoding: undefined }
What is your name - a? › jack
BEFORE - stop  { isRaw: true, isPaused: false, encoding: undefined }
AFTER  - stop  { isRaw: false, isPaused: true, encoding: undefined }

Pausing for 1 second...
BEFORE - start { isRaw: false, isPaused: true, encoding: undefined }
AFTER  - start { isRaw: true, isPaused: false, encoding: undefined }
What is your name - b? ›

TIMEOUT - exiting...
[16:43:38] stdin-bug-bun (master)$ ackb
-bash: ackb: command not found
[16:43:38] stdin-bug-bun (master)$ bun run example:bun:ignore
$ bun run stdin.js --ignore-pause
BEFORE - start { isRaw: false, isPaused: false, encoding: null }
AFTER  - start { isRaw: true, isPaused: false, encoding: undefined }
What is your name - a? › jack
BEFORE - stop  { isRaw: true, isPaused: false, encoding: undefined }
AFTER  - stop  { isRaw: false, isPaused: true, encoding: undefined }

Skipping pause...
BEFORE - start { isRaw: false, isPaused: true, encoding: undefined }
AFTER  - start { isRaw: true, isPaused: false, encoding: undefined }
What is your name - b? › jackb
BEFORE - stop  { isRaw: true, isPaused: false, encoding: undefined }
AFTER  - stop  { isRaw: false, isPaused: true, encoding: undefined }


RESULT: {
  a: "jack",
  b: "jackb",
}
[16:43:45] stdin-bug-bun (master)$ 
```

**Comment by argosphil** [comment]
I can reproduce this on `Linux 6.1.31-gentoo-dist x86_64 13th Gen Intel(R) Core(TM) i7-1370P`, and I wonder whether it's related to #8695 , which also involves stdin manipulation from an async function.

**Comment by paperclover** [comment]
this is likely the underlying cause of #7718

**Comment by argosphil** [comment]
It's this code in ProcessObjectInternals.ts:

```
  stream._read = function (size) {
    $debug("_read();", reader);
    if (!reader) {
      // TODO: this is wrong
      this.push(null);
    } else if (!shouldUnref) {
      internalRead(this);
    }
  };
```

specifically, it's the line after "TODO: this is wrong". Removing the line appears to help, but @paperdave , do you maybe remember why it's there in the first place?

Added here:
https://github.com/oven-sh/bun/commit/c8305df5c2e0e816a7efba432d4ff7b6528d43f7#diff-0b4e6adff8b4f4a0211c7babdb31d7f4de90d80ea941227d534ff2dada2296a6R192

**Comment by argosphil** [comment]
This should be fixed by #9155 . Sorry, I forgot to open a PR after pinging @paperdave .

**Comment by jackcannon** [comment]
This appears to be fixed in v1.0.30
